src/Microsoft.Repl/Parsing/CoreParseResultExtensions.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Repl.Parsing
+{
+    public static class CoreParseResultExtensions
+    {
+        public static bool ContainsExactly(this ICoreParseResult parseResult, int length, StringComparison stringComparison, params string[] sections)
+        {
+            if (parseResult.Sections.Count != length)
+            {
+                return false;
+            }
+
+            return CompareSections(parseResult.Sections, sections, stringComparison);
+        }
+
+        public static bool ContainsExactly(this ICoreParseResult parseResult, int length, params string[] sections)
+        {
+            return ContainsExactly(parseResult, length, StringComparison.OrdinalIgnoreCase, sections);
+        }
+
+        public static bool ContainsExactly(this ICoreParseResult parseResult, params string[] sections)
+        {
+            return ContainsExactly(parseResult, sections.Length, sections);
+        }
+
+        public static bool ContainsAtLeast(this ICoreParseResult parseResult, int minimumLength, StringComparison stringComparison, params string[] sections)
+        {
+            if (parseResult.Sections.Count < minimumLength)
+            {
+                return false;
+            }
+
+            return CompareSections(parseResult.Sections, sections, stringComparison);
+        }
+
+        public static bool ContainsAtLeast(this ICoreParseResult parseResult, int minimumLength, params string[] sections)
+        {
+            return ContainsAtLeast(parseResult, minimumLength, StringComparison.OrdinalIgnoreCase, sections);
+        }
+
+        public static bool ContainsAtLeast(this ICoreParseResult parseResult, params string[] sections)
+        {
+            return ContainsAtLeast(parseResult, minimumLength: sections.Length, sections);
+        }
+
+        private static bool CompareSections(IReadOnlyList<string> parseSections, string[] sections, StringComparison stringComparison)
+        {
+            for (int index = 0; index < sections.Length; index++)
```

Comment:
A single indexer over two collections is usually a recipe for an IndexOutOfRangeException.  Check for `index < parseSections.Count` as well (or `Math.Min(parseSections.Count, sections.Length)`).

Actually, this raises a question: does this compare that all sections are equal, or just that the initial N sections are equal?  Because it could also be addressed by checking whether  `parseSections.Count == sections.Length` if you're looking for an exact match.

Reply:
The fact that you're asking makes me think the methods aren't clear enough, so I might need to rethink. But to go ahead and explain: 

Probably easiest to look at the three-parameter overloads:
`ContainsExactly(this ICoreParseResult parseResult, int length, params string[] sections)` does two checks: it makes sure that parseResult has exactly `length` sections and that the first `sections.Length` of those sections matches the ones in the params arg. The common usage of this is the two-parameter overload `ContainsExactly(this ICoreParseResult parseResult, params string[] sections)` where `length` in the other overload is defaulted to `sections.Length`. In which case you're checking to make sure parseResult has exactly `sections.Length` sections and all of them match the sections in `params`.

Similarly, `ContainsAtLeast(this ICoreParseResult parseResult, int length, params string[] sections)` does two checks: it makes sure that parseResult has at least `length` sections and that the first `sections.Length` of those sections match the ones in the params arg. the common usage of this is again the two-parameter overload `ContainsAtLeast(this ICoreParseResult parseResult, params string[] sections)` where `length` in the other overload is defaulted to `sections.Length`. In which case you're checking to make sure parseResult has at least `sections.Length` sections and the first `sections.Length` of them match the sections in `params`.

The `CompareSections` function is just there to extract the duplicated logic between those two - the second check is exactly the same between the two cases, regardless of the first check. So I wrote the private function assuming that it is being called by the public functions and thus the length checks are already done there, no need to do them again. Except I didn't really do all the length checks that were necessary, so I just added those in aa1db28.